test/unit/bootstrap-specs.js

```diff
@@ -0,0 +1,98 @@
+// transpile :mocha
+
+import chai from 'chai';
+import chaiAsPromised from 'chai-as-promised';
+import 'mochawait';
+import AndroidBootstrap from '../../index';
+import ADB from 'appium-adb';
+import { withSandbox } from 'appium-test-support';
+import events from 'events';
+import UiAutomator from 'appium-uiautomator';
+import net from 'net';
+
+chai.should();
+chai.use(chaiAsPromised);
+
+describe('AndroidBootstrap', function () {
+  const systemPort = 4724;
+  let androidBootstrap = new AndroidBootstrap(systemPort),
+      adb= new ADB(),
+      uiAutomator = new UiAutomator(adb);
+
+  describe("start", withSandbox({mocks: {adb, uiAutomator, net, androidBootstrap}}, (S) => {
+    it("should return a subProcess", async function () {
+      let conn = new events.EventEmitter();
+      conn.start = () => { };
+      const appPackage = 'com.example.android.apis',
+            disableAndroidWatchers = false;
+      androidBootstrap.adb = adb;
+      androidBootstrap.uiAutomator = uiAutomator;
+      S.mocks.androidBootstrap.expects('init').once()
+        .returns('');
+      S.mocks.adb.expects('forwardPort').once()
+        .withExactArgs(systemPort, systemPort)
+        .returns('');
+      S.mocks.uiAutomator.expects("start")
+        .once()
+        .returns(conn);
+      S.mocks.net.expects('connect').once().returns(conn);
+      await androidBootstrap.start(appPackage, disableAndroidWatchers);
+      S.verify();
+    });
+  }));
+  describe("sendCommand", () => {
+    it("should timeout", async function () {
+      let conn = new events.EventEmitter();
+      conn.write = () => {};
+      conn.setEncoding = () => {};
+      androidBootstrap.socketClient = conn;
+      await androidBootstrap.sendCommand('action', {action: 'getDataDir'}, 1000)
+        .should.eventually.be.rejectedWith("Bootstrap");
+    });
+    it("should successfully return after receiving data from bootstrap in parts", async function () {
+      let conn = new events.EventEmitter();
+      conn.write = () => {};
+      conn.setEncoding = () => {};
+      androidBootstrap.socketClient = conn;
+      setTimeout(() => {
+        conn.emit("data", '{"status":0, ');
+        conn.emit("data", '"value": "hello"}');
+      }, 0);
+      await androidBootstrap.sendCommand('action', {action: 'getDataDir'}, 1000);
+    });
+    it("should successfully return after receiving data from bootstrap", async function () {
+      let conn = new events.EventEmitter();
+      conn.write = () => {};
+      conn.setEncoding = () => {};
+      androidBootstrap.socketClient = conn;
+      setTimeout(() => {
+        conn.emit("data", '{"status":0, "value": "hello"}');
+      }, 0);
+      await androidBootstrap.sendCommand('action', {action: 'getDataDir'}, 1000);
+    });
+  });
+  describe("sendAction", withSandbox({mocks: {androidBootstrap}}, (S) => {
```

Comment:
i think we should have unit tests that prove we are converting from jsonwp-style objects to normal javascript results or thrown errors